src/lib/utils/index.ts

```diff
@@ -337,27 +337,33 @@ export function prettifyHost(host: string): string {
   return host === '::' ? 'localhost' : host
 }
 
+type UnPromisify<T> = T extends Promise<infer U> ? U : T
+
 /**
  * Makes sure, that there is only one execution at a time
  * and the last invocation doesn't get lost (tail behavior of debounce)
  * Mostly designed for watch mode
  */
-export function simpleDebounce<T extends (...args: any[]) => Promise<any>>(fn: T): T {
+export function simpleDebounce<T extends (...args: any[]) => Promise<any>>(
+  fn: T
+): (...args: Parameters<T>) => { type: 'result'; data: UnPromisify<ReturnType<T>> } | { type: 'executing' } {
   let executing = false
   let pendingExecution: any = null
+  let res: any
   return (async (...args: any[]) => {
     if (executing) {
       // if there's already an execution, make it pending
       pendingExecution = args
-      return null as any
+      return { type: 'executing' }
```

Comment:
Wouldn't it be simpler to return a deferred promise here? Then the caller doesn't need to know about the inner workings between result/executing.

Reply:
What's a deferred promise in this case?

Reply:
Return a promise here, resolve it somewhere else later. (e.g. not within the promise func body)

Reply:
That won't work because promises are eager and we don't want that promise to run at all, until the one that is already running is finished. Otherwise we're not debouncing anything. We're just running the functions one after the other.

Reply:
Nevermind, that wouldn't help, because you need to dedupe higher up in the call stack.